modules/core/src/main/java/org/apache/ignite/internal/cdc/IgniteCDC.java

```diff
@@ -0,0 +1,457 @@
+import org.apache.ignite.internal.processors.cache.persistence.wal.WALPointer;
+import org.apache.ignite.internal.processors.cache.persistence.wal.reader.IgniteWalIteratorFactory;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.internal.CU;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.apache.ignite.lang.IgniteBiTuple;
+
+import static org.apache.ignite.internal.pagemem.wal.record.WALRecord.RecordType.DATA_RECORD_V2;
+import static org.apache.ignite.internal.processors.cache.persistence.wal.FileWriteAheadLogManager.WAL_NAME_PATTERN;
+
+/**
+ * CDC(Capture Data Change) application.
+ * Application run independently of Ignite node process and provide ability for the {@link CaptureDataChangeConsumer} to consume events({@link ChangeEvent}) from WAL segments.
+ * User should responsible {@link CaptureDataChangeConsumer} implementation with custom consumption logic.
+ *
+ * Ignite node should be explicitly configured for using {@link IgniteCDC}.
+ * <ol>
+ *     <li>Set {@link DataStorageConfiguration#setCdcEnabled(boolean)} to true.</li>
+ *     <li>Optional: Set {@link DataStorageConfiguration#setCdcPath(String)} to path to the directory to store WAL segments for CDC.</li>
+ *     <li>Optional: Set {@link DataStorageConfiguration#setWalForceArchiveTimeout(long)} to configure timeout for force WAL rollover,
+ *     so new events will be available for consumptions with the predicted time.</li>
+ * </ol>
+ *
+ * When {@link DataStorageConfiguration#getCdcPath()} is true then Ignite node on each WAL segment rollover creates hard link
+ * to archive WAL segment in {@link DataStorageConfiguration#getCdcPath()} directory.
+ * {@link IgniteCDC} application takes segment file and consumes events from it. After successful consumption (see {@link CaptureDataChangeConsumer#onChange(Iterator)})
+ * WAL segment will be deleted from directory.
+ *
+ * Several Ignite nodes can be started on the same host.
+ * If your deployment done with custom consistent id then you should specify it via {@link IgniteConfiguration#setConsistentId(Serializable)} in provided {@link IgniteConfiguration}.
+ *
+ * Application works as follows:
+ * <ol>
+ *     <li>Search node work directory based on provided {@link IgniteConfiguration}.</li>
+ *     <li>Await for creation of CDC directory if it not exists.</li>
+ *     <li>Acquire file lock to ensure exclusive consumption.</li>
+ *     <li>Loads state of consumption if it exists.</li>
+ *     <li>Infinetely wait for new available segment and process it.</li>
+ * </ol>
+ *
+ * @see DataStorageConfiguration#setCdcEnabled(boolean)
+ * @see DataStorageConfiguration#setCdcPath(String)
+ * @see DataStorageConfiguration#setWalForceArchiveTimeout(long)
+ * @see CommandLineStartup
+ * @see CaptureDataChangeConsumer
+ * @see DataStorageConfiguration#DFLT_CDC_PATH
+ */
+public class IgniteCDC implements Runnable {
```

Comment:
1. Should we provide a public java API to manage the app?
2. How user can gracefully shutdown the app?

Reply:
> Should we provide a public java API to manage the app?

What do you mean by "manage"?

> How user can gracefully shutdown the app?

Just `kill $pid`.
Why do we need graceful shutdown?

Reply:
> What do you mean by "manage"?

Start/stop the app.

> Just kill $pid.
> Why do we need graceful shutdown?

To make sure that all segments were processed in case of termination after a cluster was shutdown.
Ok, it can be resolved on the user side with metrics/dirs scan

Reply:
> To make sure that all segments were processed in case of termination after a cluster was shutdown.

Makes sense. 
Will implement this in code and test.

Reply:
Fixed.